File: Assets/src/model/grasses/Violets.cs
using UnityEngine;

[System.Serializable]
[ObjectInfo(spriteName: "Purple_5", description: "Alternately opens and closes every 12 turns.\nWhile open, Pacifies the creature standing over it.")]
public class Violets : Grass, ISteppable, IActorEnterHandler {
  public const int turnsToChange = 12;
  public float timeNextAction { get; set; }
  public float turnPriority => 0;
  public bool isOpen = false;
  public int countUp = 0;

  public static bool CanOccupy(Tile tile) => tile is Ground;
  public Violets(Vector2Int pos) : base(pos) {
    timeNextAction = timeCreated + 1;
    countUp = MyRandom.Range(0, 4);
  }

  public void HandleActorEnter(Actor who) {
    if (isOpen) {
      actor?.statuses.Add(new PacifiedStatus());
    }
  }

  public float Step() {
    countUp++;
    if (countUp >= turnsToChange) {
      isOpen = !isOpen;
      countUp = 0;
    }
    if (isOpen) {
      actor?.statuses.Add(new PacifiedStatus());
    }
    return 1;
  }
}

[System.Serializable]
[ObjectInfo("peace")]
public class PacifiedStatus : Status, IBaseActionModifier {
  public override string Info() => "You cannot attack while standing on an open Violet.";

  public BaseAction Modify(BaseAction input) {
    if (input.Type == ActionType.ATTACK) {
      return new StruggleBaseAction(input.actor);
    }
    return input;
  }

  public override void Step() {
    var shouldKeep = actor.grass is Violets v && v.isOpen;
    if (!shouldKeep) {
      Remove();
    }
  }

  public override bool Consume(Status other) => true;
}

File: web/src/model/grasses/Violets.ts
import { Grass } from './Grass';
import { ACTOR_ENTER_HANDLER, type IActorEnterHandler } from '../../core/types';
import { Ground } from '../Tile';
import { PacifiedStatus } from '../statuses/PacifiedStatus';
import { entityRegistry } from '../../generator/entityRegistry';
import { Actor } from '../Actor';
import type { ISteppable } from '../Floor';
import type { Vector2Int } from '../../core/Vector2Int';
import type { Tile } from '../Tile';

/**
 * Alternately opens and closes every 12 turns.
 * While open, Pacifies the creature standing over it.
 * Port of C# Violets.cs.
 */
export class Violets extends Grass implements IActorEnterHandler, ISteppable {
  readonly [ACTOR_ENTER_HANDLER] = true as const;

  static readonly turnsToChange = 12;
  timeNextAction: number;
  get turnPriority(): number { return 20; }
  isOpen = false;
  countUp: number;

  constructor(pos: Vector2Int) {
    super(pos);
    this.timeNextAction = this.timeCreated + 1;
    this.countUp = Math.floor(Math.random() * 4);
  }

  static canOccupy(tile: Tile): boolean {
    return tile instanceof Ground;
  }

  /** Actor standing on this grass's tile (null if body is not an Actor). */
  private get actor(): Actor | null {
    const body = this.floor?.bodies.get(this.pos);
    return body instanceof Actor ? body : null;
  }

  handleActorEnter(_who: any): void {
    if (this.isOpen) {
      this.actor?.statuses.add(new PacifiedStatus());
    }
  }

  step(): number {
    this.countUp++;
    if (this.countUp >= Violets.turnsToChange) {
      this.isOpen = !this.isOpen;
      this.countUp = 0;
    }
    if (this.isOpen) {
      this.actor?.statuses.add(new PacifiedStatus());
    }
    return 1;
  }
}

entityRegistry.register('Violets', Violets);
